Common/interface/LinearAllocator.hpp

```diff
@@ -0,0 +1,208 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  In no event and under no legal theory, whether in tort (including negligence), 
+ *  contract, or otherwise, unless required by applicable law (such as deliberate 
+ *  and grossly negligent acts) or agreed to in writing, shall any Contributor be
+ *  liable for any damages, including any direct, indirect, special, incidental, 
+ *  or consequential damages of any character arising as a result of this License or 
+ *  out of the use or inability to use the software (including but not limited to damages 
+ *  for loss of goodwill, work stoppage, computer failure or malfunction, or any and 
+ *  all other commercial damages or losses), even if such Contributor has been advised 
+ *  of the possibility of such damages.
+ */
+
+#pragma once
+
+/// \file
+/// Defines Diligent::LinearAllocator class
+
+#include "../../Primitives/interface/BasicTypes.h"
+#include "../../Primitives/interface/MemoryAllocator.h"
+#include "../../Platforms/Basic/interface/DebugUtilities.hpp"
+#include "Align.hpp"
+
+namespace Diligent
+{
+
+/// Implementation of a linear allocator on a fixed-size memory page
+class LinearAllocator
+{
+public:
+    // clang-format off
+    LinearAllocator           (const LinearAllocator&) = delete;
+    LinearAllocator& operator=(const LinearAllocator&) = delete;
+    LinearAllocator& operator=(LinearAllocator&&)      = delete;
+    // clang-format on
+
+    LinearAllocator(IMemoryAllocator& Allocator) :
+        m_pAllocator{&Allocator}
+    {}
+
+    LinearAllocator(LinearAllocator&& Other) :
+        // clang-format off
+        m_pBuffer     {Other.m_pBuffer     },
+        m_pCurrPtr    {Other.m_pCurrPtr    },
+        m_RequiredSize{Other.m_RequiredSize},
+        m_pAllocator  {Other.m_pAllocator  }
+    // clang-format on
+    {
+        Other.m_pBuffer      = nullptr;
+        Other.m_pCurrPtr     = nullptr;
+        Other.m_RequiredSize = 0;
+        Other.m_pAllocator   = nullptr;
+    }
+
+    ~LinearAllocator()
+    {
+        Free();
+    }
+
+    void Free()
+    {
+        if (m_pBuffer != nullptr && m_pAllocator != nullptr)
+        {
+            m_pAllocator->Free(m_pBuffer);
+        }
+
+        m_pBuffer      = nullptr;
+        m_pCurrPtr     = nullptr;
+        m_RequiredSize = 0;
+        m_pAllocator   = nullptr;
+    }
+
+    void* Release()
+    {
+        void* Ptr      = m_pBuffer;
+        m_pBuffer      = nullptr;
+        m_pCurrPtr     = nullptr;
+        m_RequiredSize = 0;
+        return Ptr;
+    }
+
+    void AddRequiredSize(size_t size, size_t align)
+    {
+        VERIFY(m_pBuffer == nullptr, "Memory already allocated");
+        if (size > 0)
+        {
+            m_RequiredSize = Align(m_RequiredSize, align) + size;
+
+            // Reserve additional space for pointer alignment
+            m_RequiredSize += (align > sizeof(void*) ? align : 0);
```

Comment:
I think this should be `+= Align(m_RequiredSize, sizeof(void*)) - m_RequiredSize` to make it aligned to the closest `sizeof(void)` rather than always reserve extra `sizeof(void*)`

Reply:
ok